core/src/main/scala/akka/kafka/javadsl/Producer.scala

```diff
@@ -46,67 +48,97 @@ object Producer {
+   * or [[ConsumerMessage.CommittableOffsetBatch CommittableOffsetBatch]] that can
+   * be committed later in the flow.
    */
-  def transactionalSink[K, V](
-    settings: ProducerSettings[K, V],
-    transactionalId: String
-  ): Sink[Message[K, V, ConsumerMessage.PartitionOffset], CompletionStage[Done]] =
-    scaladsl.Producer.transactionalSink(settings, transactionalId)
-      .mapMaterializedValue(_.toJava)
+  def flow[K, V, PassThrough](settings: ProducerSettings[K, V]): Flow[Message[K, V, PassThrough], Result[K, V, PassThrough], NotUsed] =
+    scaladsl.Producer.flow(settings)
       .asJava
+      .asInstanceOf[Flow[Message[K, V, PassThrough], Result[K, V, PassThrough], NotUsed]]
 
   /**
-   * Publish records to Kafka topics and then continue the flow.  The flow should only used with a [[Consumer]] that
-   * emits a [[ConsumerMessage.TransactionalMessage]].  The flow requires a unique `transactional.id` across all app
-   * instances.  The flow will override producer properties to enable Kafka exactly once transactional support.
-   */
-  def transactionalFlow[K, V](settings: ProducerSettings[K, V], transactionalId: String): Flow[Message[K, V, ConsumerMessage.PartitionOffset], Result[K, V, ConsumerMessage.PartitionOffset], NotUsed] =
-    scaladsl.Producer.transactionalFlow(settings, transactionalId).asJava
-
-  /**
-   * Publish records to Kafka topics and then continue the flow. Possibility to pass through a message, which
-   * can for example be a [[ConsumerMessage.CommittableOffset]] or [[ConsumerMessage.CommittableOffsetBatch]] that can
+   * Create a flow to conditionally publish records to Kafka topics and then pass it on.
+   *
+   * It publishes records to Kafka topics conditionally:
+   *
+   * - [[akka.kafka.ProducerMessage.Message Message]] publishes a single message to its topic, and continues in the stream as [[akka.kafka.ProducerMessage.Result Result]]
+   *
+   * - [[akka.kafka.ProducerMessage.MultiMessage MultiMessage]] publishes all messages in its `records` field, and continues in the stream as [[akka.kafka.ProducerMessage.MultiResult MultiResult]]
+   *
+   * - [[akka.kafka.ProducerMessage.PassThroughMessage PassThroughMessage]] does not publish anything, and continues in the stream as [[akka.kafka.ProducerMessage.PassThroughResult PassThroughResult]]
+   *
+   * The messages support the possibility to pass through arbitrary data, which can for example be a [[ConsumerMessage.CommittableOffset CommittableOffset]]
+   * or [[ConsumerMessage.CommittableOffsetBatch CommittableOffsetBatch]] that can
    * be committed later in the flow.
    */
-  def flow[K, V, PassThrough](settings: ProducerSettings[K, V]): Flow[Message[K, V, PassThrough], Result[K, V, PassThrough], NotUsed] =
-    scaladsl.Producer.flow(settings)
+  def flow2[K, V, PassThrough](settings: ProducerSettings[K, V]): Flow[Messages[K, V, PassThrough], Results[K, V, PassThrough], NotUsed] =
```

Comment:
Would it make sense to change `Message` into a class with a secondary constructor (one with a `ProducerRecord` param and one without) to keep the API code compatible and not require a second flow implementation to the new `Messages` types?

Reply:
The incompatibility is worse on the outgoing messages, as access to the pass through was available via the message field, which does not exist in the `PassThroughMessage`.
A better name than `flow2` has to be suggested, though.

Reply:
We could name this `plainFlow` which would fit the sink and source factories in this object that have the same prefix. We should also deprecate the original `flow`.

Reply:
The `plain` refers to unwrapped Kafka types, I believe.
And yes, I think the existing `flow` should become deprecated.

Reply:
To keep the convention of "_something_ Source|Sink|Flow" perhaps the new base type name could be used for `Producer.messagesFlow`.

Another idea, pass through already has meaning in this project, but what about just `through`.  `Producer.through`?